### DISCUSSION TITLE
Use with Rust

### DISCUSSION THREAD

**Discussion by joshbenaron** [taskforcesh/bullmq#988] [Ideas]
Is there a way to implement this in Rust? Or is this already done? I just want to do basic operations like add to a queue etc. Is there a rough high level api for key structures etc and what I need to store in order to do this?

**Comment by ccollie** [comment]
`bullmq` is distributed as a library, so no `api` is included, though its easy enough to set up a server with express/fastify. Check out `bull-monitor` for how they do this. They're oriented towards the UI but as part of the process they both provide an api for the frontend to use.

**Comment by ccollie** [comment]
BTW. Your use case is precisely one of the reasons I'm developing a `graphql` server on top of `bullmq`

**Reply by paulmikulskis** [reply]
thanks dope!

**Comment by joshbenaron** [comment]
Well - I might not follow. But what I want to do is use a Rust http server. For each request I add a job to the queue. I simply want to add to the queue from the Rust server

**Comment by joshbenaron** [comment]
Surely adding an item to the queue is just adding specific key/values to Redis?

**Comment by ccollie** [comment]
You can certainly add jobs via direct access to `redis`, but that's incredibly brittle. If you're really insistent and are comfortable in calling redis from rust, check out

https://github.com/taskforcesh/bullmq/blob/master/src/commands/addJob-9.lua

Calling that should get you what you want. My question though is how are your workers implemented ? I'm unaware of any non-node projects using `bullmq`. If you're using `node` workers anyway, then why not use it to add jobs ?

**Comment by joshbenaron** [comment]
So the http server is rust - and yeah the workers are nodejs. I want high performance for the http server

**Comment by joshbenaron** [comment]
>  incredibly brittle

In what sense? @ccollie

**Comment by ccollie** [comment]
Its brittle because you're sneaking around the back of the `api`.  Bullmq is implemented both in `node` and as scripts in `redis`, so it's important to know how they interact.

As far as the server being in `rust`, I honestly wouldn't worry about it, especially since all you'd be doing is pushing data to `redis`, which is IO bound in any event.  Bull also has methods to do bulk pushes if you need to squeeze out extra performance.

If you really need speed, I'd implement the workers in `rust`, since `bull` allows for manual queue handling.

**Comment by ccollie** [comment]
To be more specific, if all you're doing in the http server is pushing data to `redis`, the performance differential would not justify the introduction of a new technology, IMO. Node is perfectly at home with IO bound processing, whereas `rust` could definitely be a benefit for workers if most of your queue tasks are CPU bound.

**Comment by joshbenaron** [comment]
I meant I want performance for the server in general. One route adds to the queue. i.e. there is some CPU bound stuff that the server does

**Comment by joshbenaron** [comment]
How would I go about manually processing jobs from Rust?

**Comment by ccollie** [comment]
I guess if you have other reasons for having a `rust` based server then that's understandable.

r.e. Manual Processing:
Looking at the docs here:
https://docs.bullmq.io/patterns/manually-fetching-jobs

I may have been overly optimistic, since this uses the `bull` js library itself. It seems like you would have to replicate the worker js code in rust and follow the directions in the link.

**Comment by joshbenaron** [comment]
Cool. I'll start with the adding first. Long term I'll make this into a nice Rust lib for people to use

**Comment by ccravens** [comment]
@joshbenaron curious what performance concerns you have with NodeJS? Typically performance issues are not due to the programming language you choose but more so how the processing is implemented.

Also, if you're REALLY worried about performance have you considered another stack like ZeroMQ (https://github.com/erickt/rust-zmq) with Cap'n Proto (https://capnproto.org/). Although I'm not fully convinced that you'd likely see much in the line of performance gains to be honest.

**Reply by joshbenaron** [reply]
I want performance improvements for my http server. BullMQ isn't a bottleneck but Node.js is

**Reply by sabinthomas** [reply]
I'm in the same boat as @joshbenaron . I have multiple workers in Node.js that I am happy with, but have a new worker that is written in Rust (because there is business logic tied to a Rust library), but seem to have no way at present to be able to integrate that into my existing bull queue. It would be great for portability if there was support for this in different languages - Rust, Python, Java.

**Reply by ccollie** [reply]
My 2c - if you want to use `BullMQ`, do like I did and concentrate on the `api`. I started my project specifically for the use-case of allowing `bull` from any language/framework which supports `graphql`. You can then create minimal functionality on top of  the `graphql` client to pull jobs from the queue and update status after processing.

Unfortunately my bandwith is limited, so I'm not sure when I'll have a release

**Reply by manast** [reply]
Regarding performance, something that could be a game changer would be to port BullMQ to Bunjs: https://bun.sh/

**Reply by atrefonas** [reply]
I also want to use it with Rust, mostly for the type safety and other language semantics I prefer in Rust.

I think I am going to create a napi-rs Rust library that I can call from the node.js code. I just figured out the debugability aspect too which I had concerns about for this strategy (wasn't sure how to use a Rust debugger when starting a node.js process). But this answers it: https://github.com/napi-rs/napi-rs/discussions/722#discussioncomment-1264101

**Reply by ccollie** [reply]
Have you checked out [Neon](https://neon-bindings.com/) for `napi` ? I'm working on a lib (to be used from `BullMQ`) and will have do do the `napi-rs` thing as well

**Reply by ccollie** [reply]
Answering myself - seems like `napi-rs` is the more actively developed and ergonomic solution.

**Reply by joshbenaron** [reply]
@ccollie have you got a repo with your work? I'm working on a similar problem.

**Reply by ccollie** [reply]
@joshbenaron i'm still not through writing the lib, so the `node` interface is still a bit off,  but check out https://github.com/Brooooooklyn . He has a few projects creating `npm` packages out of `rust` libs.

**Reply by joshbenaron** [reply]
any update here @ccollie ?

**Reply by ccollie** [reply]
Not really @joshbenaron. I'm in a new position that has taken me away from `BullMQ`, though I've spent the last 8 or so months working on an opensource Rust project. I recall seeing a few projects on `crates.io` implementing workers on top of `redis`, so maybe you can look there for inspiration. There's also rustrepo.com.

If you have a project that I where I can pitch in periodically, I'd be happy to lend a hand, though my time is constrained at the moment.

**Reply by opeolluwa** [reply]
Could you please point out a few of these libraries, I need a performant Message Queue lib to connect a node and Rust application

**Comment by vstepanyuk** [comment]
Can you guide me to the documentation or code that would help me to post a `flow` (parent/children jobs) from Rust?

**Comment by manast** [comment]
@vstepanyuk you will basically need to port this class to Rust: 
https://github.com/taskforcesh/bullmq/blob/master/src/classes/flow-producer.ts
You can load the required .lua scripts from rawScripts as the python implementation is doing: 
https://github.com/taskforcesh/bullmq/blob/master/python/bullmq/scripts.py
